File: server/src/assets.ts
// Open Cloud Assets upload client (Phase 2).
//
// Uploads a local file as a Roblox asset over HTTPS. Same shape as cloud.ts
// (env-only key, create then poll an operation, transient retry), but a SEPARATE
// base path and, crucially, a multipart body whose boundary fetch must set itself,
// so this does not reuse cloud.ts's JSON fetch wrapper.
//
// Env: ROBLOX_OPEN_CLOUD_KEY (assets read + write scope) and a creator id,
// ROBLOX_CREATOR_USER_ID or ROBLOX_CREATOR_GROUP_ID. The key is sent in x-api-key
// and never logged.

import { readFileSync, statSync } from "node:fs";
import { basename, extname } from "node:path";

const ASSETS_BASE = "https://apis.roblox.com/assets/v1";
const POLL_INTERVAL_MS = 2000;
const POLL_DEADLINE_MS = 120_000;
const TRANSIENT_ATTEMPTS = 3;
const MAX_FILE_BYTES = 20 * 1024 * 1024;

export type AssetType = "Decal" | "Audio" | "Model" | "Animation" | "Video";

// The fileContent part's content type must match the asset type, so infer it from
// the extension when the caller does not pass one.
const CONTENT_TYPES: Record<string, string> = {
  ".png": "image/png",
  ".jpeg": "image/jpeg",
  ".jpg": "image/jpeg",
  ".bmp": "image/bmp",
  ".tga": "image/tga",
  ".mp3": "audio/mpeg",
  ".ogg": "audio/ogg",
  ".wav": "audio/wav",
  ".flac": "audio/flac",
  ".fbx": "model/fbx",
  ".gltf": "model/gltf+json",
  ".glb": "model/gltf-binary",
  ".rbxm": "model/x-rbxm",
  ".rbxmx": "model/x-rbxmx",
  ".mp4": "video/mp4",
  ".mov": "video/mov",
};

export interface UploadResult {
  ok: boolean;
  assetId?: string | number;
  revisionId?: string | number;
  error?: string;
}

type Credentials =
  | { ok: true; key: string; creator: { userId: string } | { groupId: string } }
  | { ok: false; error: string };

function readCredentials(): Credentials {
  const key = process.env.ROBLOX_OPEN_CLOUD_KEY;
  if (!key) return { ok: false, error: "Missing env: set ROBLOX_OPEN_CLOUD_KEY." };
  const userId = process.env.ROBLOX_CREATOR_USER_ID;
  const groupId = process.env.ROBLOX_CREATOR_GROUP_ID;
  if (userId) return { ok: true, key, creator: { userId } };
  if (groupId) return { ok: true, key, creator: { groupId } };
  return { ok: false, error: "Missing env: set ROBLOX_CREATOR_USER_ID or ROBLOX_CREATOR_GROUP_ID." };
}

export interface UploadInput {
  filePath: string;
  assetType: AssetType;
  displayName: string;
  description?: string;
  contentType?: string;
}

export async function uploadAsset(input: UploadInput): Promise<UploadResult> {
  const creds = readCredentials();
  if (!creds.ok) return { ok: false, error: creds.error };

  let fileBuffer: Buffer;
  try {
    const size = statSync(input.filePath).size;
    if (size > MAX_FILE_BYTES) {
      return { ok: false, error: `file is ${size} bytes; the per-upload limit is ${MAX_FILE_BYTES}.` };
    }
    fileBuffer = readFileSync(input.filePath);
  } catch (err) {
    return { ok: false, error: `cannot read ${input.filePath}: ${messageOf(err)}` };
  }

  const contentType = input.contentType ?? CONTENT_TYPES[extname(input.filePath).toLowerCase()];
  if (contentType === undefined) {
    return { ok: false, error: `cannot infer a content type for ${input.filePath}; pass contentType explicitly.` };
  }

  const form = new FormData();
  form.set(
    "request",
    JSON.stringify({
      assetType: input.assetType,
      displayName: input.displayName,
      description: input.description ?? "",
      creationContext: { creator: creds.creator },
    }),
  );
  // Copy into a fresh Uint8Array so the Blob part is backed by a plain ArrayBuffer.
  form.set("fileContent", new Blob([new Uint8Array(fileBuffer)], { type: contentType }), basename(input.filePath));

  let operationPath: string;
  try {
    const created = (await (
      await assetsFetch(creds.key, `${ASSETS_BASE}/assets`, { method: "POST", body: form })
    ).json()) as { path?: string };
    if (!created.path) return { ok: false, error: "create asset returned no operation path" };
    operationPath = created.path;
  } catch (err) {
    return { ok: false, error: `create asset failed: ${messageOf(err)}` };
  }

  try {
    const done = await pollOperation(creds.key, operationPath);
    // A finished operation can carry an error (for example moderation) instead of a response.
    if (done.error !== undefined) return { ok: false, error: `asset rejected: ${JSON.stringify(done.error)}` };
    return { ok: true, assetId: done.response?.assetId, revisionId: done.response?.revisionId };
  } catch (err) {
    return { ok: false, error: `polling failed: ${messageOf(err)}` };
  }
}

interface AssetOperation {
  done?: boolean;
  error?: unknown;
  response?: { assetId?: string | number; revisionId?: string | number };
}

async function pollOperation(key: string, path: string): Promise<AssetOperation> {
  const url = `${ASSETS_BASE}/${path}`;
  const deadline = Date.now() + POLL_DEADLINE_MS;
  for (;;) {
    const op = (await (await assetsFetch(key, url, { method: "GET" })).json()) as AssetOperation;
    if (op.done) return op;
    if (Date.now() > deadline) {
      throw new Error(`asset operation did not finish within ${Math.round(POLL_DEADLINE_MS / 1000)}s`);
    }
    await sleep(POLL_INTERVAL_MS);
  }
}

interface AssetsFetchInit {
  method: string;
  body?: FormData;
}

// Only adds x-api-key. It must NOT set content-type: multipart needs fetch to set
// its own boundary, and the GET poll needs no content type.
async function assetsFetch(key: string, url: string, init: AssetsFetchInit): Promise<Response> {
  let lastError = "";
  for (let attempt = 1; attempt <= TRANSIENT_ATTEMPTS; attempt++) {
    let res: Response | undefined;
    try {
      res = await fetch(url, { method: init.method, headers: { "x-api-key": key }, body: init.body });
    } catch (err) {
      lastError = messageOf(err);
    }
    if (res) {
      if (res.ok) return res;
      const detail = `${res.status} ${await safeText(res)}`.trim();
      if (!isTransient(res.status)) throw new Error(`${init.method} ${url} responded ${detail}`);
      lastError = detail;
    }
    if (attempt < TRANSIENT_ATTEMPTS) await sleep(attempt * POLL_INTERVAL_MS);
  }
  throw new Error(`${init.method} ${url} failed after ${TRANSIENT_ATTEMPTS} attempts: ${lastError}`);
}

function isTransient(status: number): boolean {
  return status === 429 || status >= 500;
}

function messageOf(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

async function safeText(res: Response): Promise<string> {
  try {
    return await res.text();
  } catch {
    return "";
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: server/src/assets.rs
// Open Cloud Assets upload client. Reads a local file, posts a multipart create to
// apis.roblox.com/assets/v1/assets, polls the operation, and returns the assetId (or
// a moderation/rejection error). The multipart boundary is set by reqwest; the key
// is sent in x-api-key only. Creator id from ROBLOX_CREATOR_USER_ID / GROUP_ID.

use std::path::Path;
use std::time::{Duration, Instant};

use reqwest::multipart::{Form, Part};
use reqwest::Client;
use serde_json::{json, Value};

use crate::env;
use crate::httpx;

const ASSETS_BASE: &str = "https://apis.roblox.com/assets/v1";
const POLL_INTERVAL: Duration = Duration::from_secs(2);
const POLL_DEADLINE: Duration = Duration::from_secs(120);
const MAX_FILE_BYTES: u64 = 20 * 1024 * 1024;

pub type UploadResult = Result<Value, String>;

fn creator() -> Result<Value, String> {
    if let Some(user) = env::var("ROBLOX_CREATOR_USER_ID") {
        Ok(json!({ "userId": user }))
    } else if let Some(group) = env::var("ROBLOX_CREATOR_GROUP_ID") {
        Ok(json!({ "groupId": group }))
    } else {
        Err("Missing env: set ROBLOX_CREATOR_USER_ID or ROBLOX_CREATOR_GROUP_ID.".into())
    }
}

fn content_type_for(path: &str, override_type: Option<&str>) -> Result<String, String> {
    if let Some(ct) = override_type {
        return Ok(ct.to_string());
    }
    let ext = Path::new(path)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_ascii_lowercase();
    let mime = match ext.as_str() {
        "png" => "image/png",
        "jpeg" | "jpg" => "image/jpeg",
        "bmp" => "image/bmp",
        "tga" => "image/tga",
        "mp3" => "audio/mpeg",
        "ogg" => "audio/ogg",
        "wav" => "audio/wav",
        "flac" => "audio/flac",
        "fbx" => "model/fbx",
        "gltf" => "model/gltf+json",
        "glb" => "model/gltf-binary",
        "rbxm" => "model/x-rbxm",
        "rbxmx" => "model/x-rbxmx",
        "mp4" => "video/mp4",
        "mov" => "video/mov",
        _ => {
            return Err(format!(
                "cannot infer a content type for {path}; pass contentType explicitly."
            ))
        }
    };
    Ok(mime.to_string())
}

pub async fn upload_asset(
    http: &Client,
    file_path: &str,
    asset_type: &str,
    display_name: &str,
    description: Option<&str>,
    content_type: Option<&str>,
) -> UploadResult {
    let key = env::var("ROBLOX_OPEN_CLOUD_KEY").ok_or("Missing env: set ROBLOX_OPEN_CLOUD_KEY.")?;
    let creator = creator()?;

    let bytes = std::fs::read(file_path).map_err(|e| format!("cannot read {file_path}: {e}"))?;
    if bytes.len() as u64 > MAX_FILE_BYTES {
        return Err(format!(
            "file is {} bytes; the per-upload limit is {MAX_FILE_BYTES}.",
            bytes.len()
        ));
    }
    let mime = content_type_for(file_path, content_type)?;
    let file_name = Path::new(file_path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("upload")
        .to_string();
    let request = serde_json::to_string(&json!({
        "assetType": asset_type,
        "displayName": display_name,
        "description": description.unwrap_or(""),
        "creationContext": { "creator": creator },
    }))
    .map_err(|e| e.to_string())?;

    let created = httpx::send_retrying(|| {
        // Built fresh each attempt: a multipart body is consumed on send and the
        // bytes are read in once, so cloning here keeps retries replayable.
        let request_part = Part::text(request.clone())
            .mime_str("application/json")
            .expect("application/json is a valid mime");
        let file_part = Part::bytes(bytes.clone())
            .file_name(file_name.clone())
            .mime_str(&mime)
            .expect("inferred mime is valid");
        let form = Form::new()
            .part("request", request_part)
            .part("fileContent", file_part);
        http.post(format!("{ASSETS_BASE}/assets"))
            .header("x-api-key", &key)
            .multipart(form)
    })
    .await
    .map_err(|e| format!("create asset failed: {e}"))?;

    let operation: Value = created.json().await.map_err(|e| e.to_string())?;
    let op_path = operation
        .get("path")
        .and_then(Value::as_str)
        .ok_or("create asset returned no operation path")?
        .to_string();

    let done = poll_operation(http, &key, &op_path).await?;
    if let Some(error) = done.get("error") {
        if !error.is_null() {
            return Err(format!("asset rejected: {error}"));
        }
    }
    let response = done.get("response").cloned().unwrap_or(Value::Null);
    Ok(json!({
        "assetId": response.get("assetId"),
        "revisionId": response.get("revisionId"),
    }))
}

async fn poll_operation(http: &Client, key: &str, path: &str) -> Result<Value, String> {
    let url = format!("{ASSETS_BASE}/{path}");
    let deadline = Instant::now() + POLL_DEADLINE;
    loop {
        let op = httpx::send_retrying(|| http.get(&url).header("x-api-key", key)).await?;
        let body: Value = op.json().await.map_err(|e| e.to_string())?;
        if body.get("done").and_then(Value::as_bool).unwrap_or(false) {
            return Ok(body);
        }
        if Instant::now() > deadline {
            return Err(format!(
                "asset operation did not finish within {}s",
                POLL_DEADLINE.as_secs()
            ));
        }
        tokio::time::sleep(POLL_INTERVAL).await;
    }
}
